### PR TITLE
fix Issue 18057 - [ICE] Segmentation fault (stack overflow) in Expression::ctfeInterpret()

### PR DESCRIPTION
**Pull request by ibuclaw**
Inside `AggregateDeclaration::fill` looks to be the better place to have this error rather than `VarDeclaration::getConstInitializer`.  In the latter location you'd get the same error printed twice.

### CODE REVIEW

**Review by JinShil — APPROVED** [review]

File: test/compilable/interpret3.d [hunk 1]
```diff
@@ -7731,3 +7731,14 @@ bool foo17407()
...
+/**************************************************/
+// https://issues.dlang.org/show_bug.cgi?id=18057
+// Recursive field initializer causes segfault.
+
+struct RBNode(T)
+{
+    RBNode!T *copy = new RBNode!T;
+}
+
+static assert(!__traits(compiles, { alias bug18057 = RBNode!int; }));
```
Comment (wilzbach) [inline test/compilable/interpret3.d]: Shouldn't we test for the exact error message here?

Comment (ibuclaw) [reply test/compilable/interpret3.d]: If you insist.  Adding more files to the testsuite has been complained about in the past, as it is gradually slowing down the time it takes to run the testsuite vs. if all tests were kept in one mega file.  (Not saying everything should go in one mega file, but logically grouped may be a start :-)

Comment (ibuclaw) [reply test/compilable/interpret3.d]: I created an issue to note this, incase anyone wants to take the baton: https://issues.dlang.org/show_bug.cgi?id=18235

Comment (WalterBright) [reply test/compilable/interpret3.d]: The exact error message only needs to be tested once by the test suite, the others can use `__traits(compiles, ... )`. So, the coverage results can be used to determine if the exact message is ever tested or not.

File: test/compilable/interpret3.d [hunk 2]
```diff
@@ -1,4 +1,4 @@
-﻿// PERMUTE_ARGS: -inline
+// PERMUTE_ARGS: -inline
```
Comment (ibuclaw) [inline test/compilable/interpret3.d]: What's going on with this?

**Comment by ibuclaw** [comment]
Gah, I was just about to do this. :-)

File: test/fail_compilation/fail18057.d [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+TEST_OUTPUT:
+---
+fail_compilation/fail18057.d(13): Error: variable fail18057.RBNode!int.RBNode.copy recursive initialization of field
+fail_compilation/fail18057.d(16): Error: template instance fail18057.RBNode!int error instantiating
+---
+*/
+
+// https://issues.dlang.org/show_bug.cgi?id=18057
+// Recursive field initializer causes segfault.
+struct RBNode(T)
```
Comment (ibuclaw) [inline test/fail_compilation/fail18057.d]: Doesn't need to be a template.  Only did that for the benefit of `__traits(compiles)`.

File: test/compilable/interpret3.d (hunk 1 above)
Comment (wilzbach) [reply test/compilable/interpret3.d]: I just moved it to `fail_compilation`. It's just that adding `static assert(!__traits(compiles,` to `compilable` tests strikes me as an anti-pattern

File: test/compilable/interpret3.d (hunk 2 above)
Comment (wilzbach) [reply test/compilable/interpret3.d]: Trailing whitespace - my editor automatically removes it. Sorry

File: test/fail_compilation/fail18057.d (hunk 3 above)
Comment (wilzbach) [reply test/fail_compilation/fail18057.d]: OK. Removed.

**Pull request merged**